Exercises/advanced_reductions/Begin/advanced_reductions.cpp

```diff
@@ -0,0 +1,56 @@
+//@HEADER
+// ************************************************************************
+//
+//                        Kokkos v. 4.0
+//       Copyright (2022) National Technology & Engineering
+//               Solutions of Sandia, LLC (NTESS).
+//
+// Under the terms of Contract DE-NA0003525 with NTESS,
+// the U.S. Government retains certain rights in this software.
+//
+// Part of Kokkos, under the Apache License v2.0 with LLVM Exceptions.
+// See https://kokkos.org/LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//@HEADER
+
+#include <Kokkos_Core.hpp>
+#include <iostream>
+
+template <typename ValueType>
+struct GeometricMean {
+  using value_type = ValueType;
+
+  GeometricMean(Kokkos::View<value_type *> view)
+      : m_view(view), n(view.size()) {}
+
+  /* Fill class */
+
+private:
+  Kokkos::View<value_type *> m_view;
+  int n;
+};
+
+int main(int argc, char *argv[]) {
+  Kokkos::initialize(argc, argv);
+  {
+    int n = 10;
+    Kokkos::View<double *> view("view", n);
+    Kokkos::parallel_for(
+        n, KOKKOS_LAMBDA(int i) { view(i) = 1 + i / 10.; });
+
+    double result;
+    Kokkos::parallel_reduce(n, GeometricMean{view}, result);
```

Comment:
I think it would have been better to write a classic C++ host only implementation with comments saying to initialize on the device etc. rather than having code that give a compile error.

Also (assuming you wrote host only code) without comments clearly indicating you want a functor with join/init/final one could propose a regular reduce with a product reducer and a lambda.

If your exercise is about writing special functors, then maybe elaborate in comments to clearly express the intent.

Reply:
> Also (assuming you wrote host only code) without comments clearly indicating you want a functor with join/init/final one could propose a regular reduce with a product reducer and a lambda.

Yeah, the point is that you can only take the n-th root at the end with a custom reducer.

Reply:
Precomputing the normalization value and doing it as part of the operator() is likely the more efficient implementation.  My point is it is ok to do something a bit contrived as an exercise as long as your add enough comments